tron/eventbus.py

```diff
@@ -0,0 +1,162 @@
+import logging
+import os
+import pickle
+import time
+from collections import deque
+
+from twisted.internet import reactor
+
+log = logging.getLogger(__name__)
+
+
+def make_eventbus(log_dir):
+    return EventBus(log_dir)
+
+
+def consume_dequeue(queue, func):
+    queue_length = len(queue)
+    for _ in range(queue_length):
+        func(queue.popleft())
+
+
+class EventBus:
+    def __init__(self, log_dir):
+        self.must_shutdown = False
+        self.event_log = {}
+        self.event_subscribers = {}
+        self.publish_queue = deque()
+        self.subscribe_queue = deque()
+        self.log_dir = log_dir
+        self.log_current = os.path.join(self.log_dir, "current")
+        self.log_updates = 0
+        self.log_last_save = 0
+        self.log_save_interval = 60
+        self.log_save_updates = 10
+
+    def start(self):
+        log.info("starting")
+        self.setup_log_dir()
+        self.sync_load_log()
+        reactor.callLater(0, self.sync_loop)
+
+    def setup_log_dir(self):
+        """Create log directory and link to current log if those don't
+        already exist"""
+        if not os.path.exists(self.log_dir):
+            log.warning(f"creating {self.log_dir}")
+            os.mkdir(self.log_dir)
+
+        if not os.path.exists(self.log_current) or not os.path.exists(
+            os.readlink(self.log_current)
+        ):
+            log.warning(f"creating {self.log_current}")
+            self.sync_save_log("initial save")
+
+    def shutdown(self):
+        self.must_shutdown = True
+        log.info(f"shutdown requested")
+
+    def publish(self, event):
+        self.publish_queue.append(event)
+        log.debug(f"publish of {event['id']} enqueued")
+
+    def subscribe(self, prefix, subscriber, callback):
+        self.subscribe_queue.append((prefix, subscriber, callback))
+        log.debug(f"subscription ({prefix}, {subscriber}) enqueued")
+
+    def has_event(self, event_id):
+        return event_id in self.event_log
+
+    def sync_load_log(self):
+        started = time.time()
+        with open(self.log_current, 'rb') as f:
+            self.event_log = pickle.load(f)
+        duration = time.time() - started
+        log.info(f"log read from disk, took {duration:.4}s")
+
+    def sync_save_log(self, reason):
+        started = time.time()
+        new_file = os.path.join(self.log_dir, f"{int(started)}.pickle")
+        with open(new_file, 'xb') as f:
+            pickle.dump(self.event_log, f)
```

Comment:
Do we have to pickle? This decision could hurt us in the long run. Sure we can't get away with good ole json?

Reply:
json is orders of magnitude slower. i think we will be fine - we never rewrite these files, we store all history of backups (i'll also ship a cron that will implement some sensible retention policy) and most importantly - what goes into these files are dumb dicts of strings to strings (not intended to change). this is as safe as you can be with pickle.